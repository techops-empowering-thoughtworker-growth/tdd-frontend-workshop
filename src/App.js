import React, { useEffect, useState } from 'react';
import './styles/_style.scss';
import axios from 'axios';

function App() {
  const [todoItems, setTodoItems] = useState([]);
  const [singleItem, setSingleItem] = useState('');

  const saveTask = () => {
    axios
      .post('http://localhost:8080/tasks', { name: singleItem });
      todoItems.push({
          id: 3,
          name: 'name3',
          completed: false
      });
      console.log(todoItems);
      setTodoItems(todoItems);
  };

  useEffect(
    () =>
      axios
        .get('http://localhost:8080/tasks')
        .then((res) => res.data)
        .then((items) => setTodoItems(items)),
    []
  );

  return (
    <div className={'main'}>
      <div className={'input-component'}>
        <h1 className={'title'}>Use this to manage your work and life</h1>
        <div className={'input-bar'}>
          <input
            className={'add-input'}
            placeholder={'Enter your todo item'}
            onChange={(event) => setSingleItem(event.target.value)}
          />
          <div className={'add-button'} onClick={saveTask}>
            +
          </div>
        </div>
      </div>
      <div className={'todo-list'}>
        {todoItems
          .filter((item) => !item.completed)
          .map((item) => (
            <li className={'task-item'} key={item.id}>
              <input type={'checkbox'} />
              <p className={'task-text'}>{item.name}</p>
            </li>
          ))}
      </div>
      <div className={'completed-list'}>
        <h2 className={'completed-title'}>Completed</h2>
        {todoItems
          .filter((item) => item.completed)
          .map((item) => (
            <li className={'task-item'} key={item.id}>
              <input type={'checkbox'} />
              <p className={'task-text'}>{item.name}</p>
            </li>
          ))}
      </div>
    </div>
  );
}

export default App;
