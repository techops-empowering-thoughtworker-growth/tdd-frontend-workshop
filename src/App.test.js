import App from './App';
import React from 'react';
import axios from 'axios';
import { mount } from 'enzyme';
import { render, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

const getResponse = {
  data: [
    {
      id: 1,
      name: 'task 01',
      completed: false,
    },
    {
      id: 2,
      name: 'task 02',
      completed: true,
    },
  ],
};

const addResponse = {
  data: {
    id: 3,
    name: 'task 03',
    completed: false,
  },
};

const checkResponse = {
  data: {
    id: 1,
    name: 'task 01',
    completed: true,
  },
};

const uncheckResponse = {
  data: {
    id: 2,
    name: 'task 02',
    completed: false,
  },
};

jest.mock('axios');

describe('App', () => {
  beforeEach(() => {
    axios.get.mockResolvedValue(getResponse);
  });

  it('should list correctly using testLibrary', async () => {
    const mockAxios = axios.get.mockResolvedValue(getResponse);
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText('task 01')).toBeInTheDocument();
      expect(screen.getByText('task 02')).toBeInTheDocument();
    });
    expect(mockAxios).toHaveBeenCalledWith('http://localhost:8080/tasks');
  });

  it('should list correctly using Enzyme and Jest', async () => {
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    wrapper.update();
    expect(axios.get).toHaveBeenCalledWith('http://localhost:8080/tasks');
    expect(wrapper.find('.task-item')).toHaveLength(2);
    expect(wrapper.find('.task-item').at(0).find('.task-text').text()).toBe(
      'task 01'
    );
    expect(wrapper.find('.task-item').at(1).find('.task-text').text()).toBe(
      'task 02'
    );
  });

  it('should add correctly', async () => {
    let wrapper;
    const mockAdd = axios.post.mockResolvedValue(addResponse);

    await act(async () => {
      wrapper = mount(<App />);
    });
    wrapper.update();

    wrapper
      .find('.add-input')
      .simulate('change', { target: { value: 'task 03' } });

    await act(async () => {
      wrapper.find('.add-button').simulate('click');
    });
    wrapper.update();

    expect(mockAdd).toHaveBeenCalledWith('http://localhost:8080/tasks', {
      completed: false,
      name: 'task 03',
    });
    expect(wrapper.find('.task-item')).toHaveLength(3);
  });

  it('should delete correctly', async () => {
    let wrapper;
    axios.delete.mockResolvedValue();

    await act(async () => {
      wrapper = mount(<App />);
    });
    wrapper.update();

    await act(async () => {
      wrapper
        .find('.task-item')
        .at(0)
        .simulate('mouseover')
        .find('.bi-trash')
        .simulate('click');
    });
    wrapper.update();

    expect(axios.delete).toHaveBeenCalledWith('http://localhost:8080/tasks/1');
    expect(wrapper.find('.task-item')).toHaveLength(1);
    expect(wrapper.find('.task-item').find('.task-text').text()).toBe(
      'task 02'
    );
  });

  it('should check task', async () => {
    let wrapper;
    axios.put.mockResolvedValue(checkResponse);

    await act(async () => {
      wrapper = mount(<App />);
    });
    wrapper.update();

    await act(async () => {
      wrapper
        .find('.task-item')
        .at(0)
        .find({ type: 'checkbox' })
        .simulate('click');
    });
    wrapper.update();

    expect(axios.put).toHaveBeenCalledWith('http://localhost:8080/tasks/1', {
      name: 'task 01',
      completed: true,
    });
    expect(wrapper.find('.completed-list').find('.task-item')).toHaveLength(2);
  });

  it('should uncheck task', async () => {
    let wrapper;
    axios.put.mockResolvedValue(uncheckResponse);

    await act(async () => {
      wrapper = mount(<App />);
    });
    wrapper.update();

    await act(async () => {
      wrapper
        .find('.task-item')
        .at(1)
        .find({ type: 'checkbox' })
        .simulate('click');
    });
    wrapper.update();

    expect(axios.put).toHaveBeenCalledWith('http://localhost:8080/tasks/2', {
      name: 'task 02',
      completed: false,
    });
    expect(wrapper.find('.todo-list').find('.task-item')).toHaveLength(2);
  });
});
