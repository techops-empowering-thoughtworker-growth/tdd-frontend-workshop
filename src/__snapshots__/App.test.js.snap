// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
LoadedCheerio {
  "0": <div
    class="main"
  >
    <div
      class="input-component"
    >
      <h1
        class="title"
      >
        Use this to manage your life and work
      </h1>
      <div
        class="input-bar"
      >
        <input
          class="add-input"
          placeholder="Enter your todo item"
          value=""
        />
        <div
          class="add-button"
        >
          +
        </div>
      </div>
    </div>
    <div
      class="todo-list"
    />
    <div
      class="completed-list"
    >
      <h2>
        Completed
      </h2>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
